### PR TITLE
WT-7400 Set WT_CURSTD_HS_READ_ALL flag for the search before fixing the out of order timestamps

### PR DESCRIPTION
**Pull request by mm-ng**

### CODE REVIEW

File: src/history/hs_rec.c [hunk 1]
```diff
@@ -175,10 +175,16 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
         if (!hs_read_all_flag)
             F_CLR(cursor, WT_CURSTD_HS_READ_ALL);
     }
-    if (ret == 0)
+    if (ret == 0) {
```
Comment (quchenhao) [inline src/history/hs_rec.c:178]: The flag is set at the wrong place. It should be on the previous two search near `__wt_curhs_search_near_after`

**Review by quchenhao — CHANGES_REQUESTED** [review]
The flag should be set at the search near calls.

**Review by quchenhao — APPROVED** [review]

File: src/history/hs_rec.c [hunk 2]
```diff
@@ -668,20 +668,26 @@ __wt_hs_delete_key_from_ts(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint3
...
-    bool hs_read_committed;
+    bool hs_read_all_flag, hs_read_committed;
 
     /*
      * If we will delete all the updates of the key from the history store, we should not reinsert
      * any update.
      */
     WT_ASSERT(session, ts > WT_TS_NONE || !reinsert);
 
+    hs_read_all_flag = F_ISSET(hs_cursor, WT_CURSTD_HS_READ_ALL);
```
Comment (quchenhao) [inline src/history/hs_rec.c:679]: No need to set WT_CURSTD_HS_READ_COMMITTED flag if we set WT_CURSTD_HS_READ_ALL.

Comment (quchenhao) [reply src/history/hs_rec.c:679]: Can you run a mongodb patch build?

File: src/history/hs_rec.c [hunk 3]
```diff
@@ -668,20 +668,26 @@ __wt_hs_delete_key_from_ts(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint3
...
     if (!hs_read_committed)
         F_SET(hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
 
     hs_cursor->set_key(hs_cursor, 3, btree_id, key, ts);
+    F_SET(hs_cursor, WT_CURSTD_HS_READ_ALL);
     WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_after(session, hs_cursor), true);
+
+    if (!hs_read_all_flag)
+        F_CLR(hs_cursor, WT_CURSTD_HS_READ_ALL);
+
```
Comment (quchenhao) [inline src/history/hs_rec.c:690]: We need to clear the flag in err as well.

File: src/history/hs_rec.c [hunk 4]
```diff
@@ -668,20 +668,20 @@ __wt_hs_delete_key_from_ts(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint3
...
 
-    hs_read_committed = F_ISSET(hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
-    if (!hs_read_committed)
-        F_SET(hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
+    hs_read_all_flag = F_ISSET(hs_cursor, WT_CURSTD_HS_READ_ALL);
 
     hs_cursor->set_key(hs_cursor, 3, btree_id, key, ts);
+    F_SET(hs_cursor, WT_CURSTD_HS_READ_ALL);
     WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_after(session, hs_cursor), true);
+
```
Comment (kommiharibabu) [inline src/history/hs_rec.c:684]: No need for this new line.

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Pull request merged**